File: 2024/q08.py
#!/usr/bin/env python
import itertools


def part1(data):
    n = int(data)
    i = int(n ** 0.5)
    if i * i < n:
        i += 1
    ans = (i * 2 - 1) * (i * i - n)
    return ans


def part2(data, M=1111, N=20240000):
    n = int(data)
    s = 1
    t = 1
    for i in itertools.count(2):
        t = (t * n) % M
        s += (i * 2 - 1) * t
        if s >= N:
            break
    ans = (i * 2 - 1) * (s - N)
    return ans


def part3(data, M=10, N=202400000):
    n = int(data)
    t = 1
    f = [1]
    for i in itertools.count(2):
        t = (t * n) % M + M
        f = [x+t for x in f] + [t]
        w = i * 2 - 1
        g = [x - ((n * w) * x % M) for x in f[:-1]] + f[-1:]
        s = sum(g) + sum(g[1:])
        if s >= N:
            break
    ans = s - N
    return ans


data = '13'
assert part1(data) == 21

data = '3'
assert part2(data, M=5, N=50) == 27

data = '2'
assert part3(data, M=5, N=160) == 2


data = open('q08_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q08_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q08_p3.txt').read()
ans = part3(data)
print('part3:', ans)

File: 2024/q08.jl
#!/usr/bin/env julia

function part1(data)
    n = parse(Int, data)
    i = isqrt(n)+1
    (2i-1) * (i*i - n)
end


function part2(data, M=1111, N=20240000)
    n = parse(Int, data)
    t, w, r = 1, 1, 1
    while r <= N
        t += 2
        w = w * n % M
        r += t * w
    end
    t * (r - N)
end


function part3(data, M=10, N=202400000)
    n = parse(Int, data)
    t, w, r, s = 1, 1, 1, 1
    h = [1]
    while s <= N
        t += 2
        w = w * n % M + M
        r += t * w
        push!(h, 0)
        h .+= w
        x = 2sum(t*n*i%M for i in h[1:end-1]) - (t*n*h[1]%M)
        s = r - x
    end
    return (s - N)
end


@assert part1("13") == 21
@assert part2("3", 5, 50) == 27
@assert part3("2", 5, 160) == 2


data = readchomp("q08_p1.txt")
println("part1: ", part1(data))
data = readchomp("q08_p2.txt")
println("part2: ", part2(data))
data = readchomp("q08_p3.txt")
println("part3: ", part3(data))
